### ISSUE TITLE
When to stop training | Major increase after 400,000 iterations

### ISSUE THREAD

**Issue by dexception** [AlexeyAB/darknet#2657]
I have been checking the output constantly and there seems to major improvements in the prediction if i allow the model to be trained for like 400,000 Iterations.
![chart](https://user-images.githubusercontent.com/31848222/54623491-9ca0ad80-4a91-11e9-8899-f4cfbd995e2b.png)

And this dataset contains only 5K images. I have the following queries:

1. How long should i further train for maximum accuracy ?
2. Even after the training is automatically finished can i still improve the accuracy ? If so do i need to create another dataset on top of this ?

Thanks.

**Comment by wonchulSon** [comment]
@dexception 
Did you use COCO dataset?
And what was your exact command?

**Comment by dexception** [comment]
@wonchulSon 
> This is my own dataset that have tagged.

> Command
./darknet detector train custom.data yolov3-tiny.cfg yolov3-tiny.conv.15

Training finished After 500K
Now i am again fine tuning it on the same dataset with the partial command. This is where i am confused should i be training on the same dataset ? or tag more data ? Since there is scope for improvement.

**Comment by AlexeyAB** [comment]
> I have been checking the output constantly and there seems to major **improvements in the prediction** if i allow the model to be trained for like 400,000 Iterations.

Did you check mAP on training or validation dataset?

----

> How long should i further train for maximum accuracy ?

Train while increasing accuracy mAP: https://github.com/AlexeyAB/darknet#when-should-i-stop-training

----

> Even after the training is automatically finished can i still improve the accuracy ? If so do i need to create another dataset on top of this ?

You can increase `max_batches=` in your cfg-file and continue training.

**Comment by dexception** [comment]
```
 43877: 0.311875, 0.461892 avg loss, 0.001000 rate, 0.617012 seconds, 2808128 images
Loaded: 0.000022 seconds
Region 16 Avg IOU: 0.852992, Class: 0.999993, Obj: 0.999215, No Obj: 0.000378, .5R: 1.000000, .75R: 1.000000,  count: 1
Region 23 Avg IOU: 0.852512, Class: 0.999898, Obj: 0.894832, No Obj: 0.000825, .5R: 1.000000, .75R: 0.933333,  count: 15
Region 16 Avg IOU: 0.859232, Class: 0.999833, Obj: 0.841465, No Obj: 0.001586, .5R: 1.000000, .75R: 1.000000,  count: 6
Region 23 Avg IOU: 0.837402, Class: 0.999926, Obj: 0.843234, No Obj: 0.000487, .5R: 1.000000, .75R: 0.909091,  count: 11
Region 16 Avg IOU: 0.743024, Class: 0.999858, Obj: 0.927524, No Obj: 0.000598, .5R: 1.000000, .75R: 0.500000,  count: 2
Region 23 Avg IOU: 0.815899, Class: 0.999819, Obj: 0.888392, No Obj: 0.002878, .5R: 1.000000, .75R: 0.825397,  count: 63
Region 16 Avg IOU: 0.896243, Class: 0.999920, Obj: 0.970498, No Obj: 0.000837, .5R: 1.000000, .75R: 1.000000,  count: 3
Region 23 Avg IOU: 0.719276, Class: 0.999305, Obj: 0.727393, No Obj: 0.001945, .5R: 0.913043, .75R: 0.521739,  count: 46
Region 16 Avg IOU: 0.871533, Class: 0.999654, Obj: 0.989912, No Obj: 0.000733, .5R: 1.000000, .75R: 1.000000,  count: 3
Region 23 Avg IOU: 0.818077, Class: 0.999899, Obj: 0.920456, No Obj: 0.003161, .5R: 0.954545, .75R: 0.878788,  count: 66
Region 16 Avg IOU: 0.864250, Class: 0.999271, Obj: 0.955767, No Obj: 0.002379, .5R: 1.000000, .75R: 1.000000,  count: 5
Region 23 Avg IOU: 0.737647, Class: 0.999718, Obj: 0.871095, No Obj: 0.002203, .5R: 0.954545, .75R: 0.568182,  count: 44
Region 16 Avg IOU: 0.892284, Class: 0.999793, Obj: 0.932264, No Obj: 0.001376, .5R: 1.000000, .75R: 1.000000,  count: 6
Region 23 Avg IOU: 0.795495, Class: 0.999775, Obj: 0.852506, No Obj: 0.000640, .5R: 1.000000, .75R: 0.769231,  count: 13
Region 16 Avg IOU: 0.865016, Class: 0.999983, Obj: 0.998657, No Obj: 0.000257, .5R: 1.000000, .75R: 1.000000,  count: 1
Region 23 Avg IOU: 0.797990, Class: 0.999617, Obj: 0.700916, No Obj: 0.000613, .5R: 1.000000, .75R: 0.866667,  count: 15
```

I am retraining on the same dataset after extracting the first 15 layers.

I have an interesting use case where i am only looking to detect faces with -60 to +60 degrees.
So all the images have been carefully tagged. If i continue retraining on the same dataset again and again will this lead to increase in accuracy or i need to tag more images ?

**Comment by AlexeyAB** [comment]
If you will continue retraining on the same dataset again and again 

* this will lead to increase in accuracy on Training dataset
* and this will lead to decrease in accuracy on Valid/Test datasets

This is **overfitting**: https://github.com/AlexeyAB/darknet#when-should-i-stop-training

**Comment by dexception** [comment]
@AlexeyAB 
Thanks. I will add more images and retrain to fill the missing gaps.
Thanks for your support. 

Waiting for the yolo2_light c++ code.
